From 96c7b0a2f25313a694c1d255abb94b056fde2626 Mon Sep 17 00:00:00 2001
From: Pardha Kesapragada <pkesapragada@habana.ai>
Date: Tue, 23 Feb 2021 09:40:07 +0200
Subject: [PATCH] [SW-20154] pcie: change iatu configuration

For Gaudi2, IATU configuration differs from Goya2.
Hence make the configuration function platform
specific such that Gaudi2 can confgure a different
content.

Change-Id: I5ee732612a1fa62bea8e90d912d48573f00ef431
Signed-off-by: Pardha Kesapragada <pkesapragada@habana.ai>
---

diff --git a/common/pcie.c b/common/pcie.c
index d23174c..2502590 100644
--- a/common/pcie.c
+++ b/common/pcie.c
@@ -162,11 +162,7 @@
 		}
 
 		/* Configure IATU */
-		regmap_writel(0x300000, mmPCIE_AUX_DBI);
-		regmap_writel(0xF0000000, IATU_LWR_TARGET_ADDR_OFF_INBOUND);
-		regmap_writel(0x7F, IATU_UPPER_TARGET_ADDR_OFF_INBOUND);
-		regmap_writel(0, IATU_REGION_CTRL_1_OFF_INBOUND);
-		regmap_writel(0xC0080000, IATU_REGION_CTRL_2_OFF_INBOUND);
+		pcie_config_iatu();
 
 		/* enable flr handling by HW state machine */
 		enable_flr_handling_fsm();
diff --git a/common/pcie.h b/common/pcie.h
index 7475675..8cf83fe 100644
--- a/common/pcie.h
+++ b/common/pcie.h
@@ -12,16 +12,11 @@
 #define PCIE_PRSTN_ASSERT			0
 #define PCIE_PRSTN_DEASSERT			1
 
-#define IATU_BASE				(mmPCIE_DBI_BASE & 0xFFFFFF)
-#define IATU_REGION_CTRL_1_OFF_INBOUND		(IATU_BASE + 0x100)
-#define IATU_REGION_CTRL_2_OFF_INBOUND		(IATU_BASE + 0x104)
-#define IATU_LWR_TARGET_ADDR_OFF_INBOUND	(IATU_BASE + 0x114)
-#define IATU_UPPER_TARGET_ADDR_OFF_INBOUND	(IATU_BASE + 0x118)
-
 #define PCIE_PHY_PROTOCOL0_EXT_RX_ADAPT_MODE_G3_CONF	0x5555
 #define PCIE_PHY_PROTOCOL0_EXT_RX_ADAPT_SEL_G4_CONF	0xFF
 
 void pcie_recovery(void);
 void execute_pcie_recovery(void);
+void pcie_config_iatu(void);
 
 #endif /* _PCIE_H_ */
diff --git a/gaudi2_arc_gnu/common.c b/gaudi2_arc_gnu/common.c
index 28c8ef0..5558292 100644
--- a/gaudi2_arc_gnu/common.c
+++ b/gaudi2_arc_gnu/common.c
@@ -25,6 +25,16 @@
 #define SP_SRAM_OFFS_MASK		0xFFFF
 #define SRAM_PHYS_OFFS_MASK		0x2FFFFFF
 
+#define IATU_REGION_CTRL_1_OFF_INBOUND_OFF	0x100
+#define IATU_REGION_CTRL_2_OFF_INBOUND_OFF	0x104
+#define IATU_LWR_TARGET_ADDR_OFF_INBOUND_OFF	0x114
+#define IATU_UPPER_TARGET_ADDR_OFF_INBOUND_OFF	0x118
+
+#define IATU_LOWER_TARGET_ADDR			0xF4000000
+#define IATU_HIGHER_TARGET_ADDR			0x1000007F
+#define IATU_INBOUND_LOW_OFFSET_ADDR		0x0
+#define IATU_INBOUND_HI_OFFSET_ADDR		0xC0080000
+#define DBI_ENABLE				0x300000
 
 void arcboot_rom_status_update(enum arcboot_rom_progress progress)
 {
@@ -325,3 +335,18 @@
 	/* limit flash write to just 2MB */
 	regmap_writel(write_en_2MB, mmPSOC_SECURITY_SPI_WRITE_LIMIT);
 }
+
+void pcie_config_iatu(void)
+{
+	uint32_t pci_dbi_base = (mmPCIE_DBI_BASE & 0x7FFFFFF);
+
+	regmap_writel(DBI_ENABLE, mmPCIE_AUX_DBI);
+	regmap_writel(IATU_LOWER_TARGET_ADDR,
+		(pci_dbi_base + IATU_LWR_TARGET_ADDR_OFF_INBOUND_OFF));
+	regmap_writel(IATU_HIGHER_TARGET_ADDR,
+		(pci_dbi_base + IATU_UPPER_TARGET_ADDR_OFF_INBOUND_OFF));
+	regmap_writel(IATU_INBOUND_LOW_OFFSET_ADDR,
+		(pci_dbi_base + IATU_REGION_CTRL_1_OFF_INBOUND_OFF));
+	regmap_writel(IATU_INBOUND_HI_OFFSET_ADDR,
+		(pci_dbi_base + IATU_REGION_CTRL_2_OFF_INBOUND_OFF));
+}
diff --git a/goya2_arc/common.c b/goya2_arc/common.c
index 6109d26..fa82171 100644
--- a/goya2_arc/common.c
+++ b/goya2_arc/common.c
@@ -11,6 +11,12 @@
 #include <psoc_global_conf_masks.h>
 #include <arcboot.h>
 
+#define IATU_BASE				(mmPCIE_DBI_BASE & 0xFFFFFF)
+#define IATU_REGION_CTRL_1_OFF_INBOUND		(IATU_BASE + 0x100)
+#define IATU_REGION_CTRL_2_OFF_INBOUND		(IATU_BASE + 0x104)
+#define IATU_LWR_TARGET_ADDR_OFF_INBOUND	(IATU_BASE + 0x114)
+#define IATU_UPPER_TARGET_ADDR_OFF_INBOUND	(IATU_BASE + 0x118)
+
 uint32_t get_efuse_read_reg(void)
 {
 	return mmPSOC_EFUSE_READ;
@@ -51,3 +57,12 @@
 	flr_ctrl |= (1 << PCIE_AUX_FLR_CTRL_FSM_EN_SHIFT);
 	regmap_writel(flr_ctrl, mmPCIE_AUX_FLR_CTRL);
 }
+
+void pcie_config_iatu(void)
+{
+	regmap_writel(0x300000, mmPCIE_AUX_DBI);
+	regmap_writel(0xF0000000, IATU_LWR_TARGET_ADDR_OFF_INBOUND);
+	regmap_writel(0x7F, IATU_UPPER_TARGET_ADDR_OFF_INBOUND);
+	regmap_writel(0, IATU_REGION_CTRL_1_OFF_INBOUND);
+	regmap_writel(0xC0080000, IATU_REGION_CTRL_2_OFF_INBOUND);
+}
